src/libraries/System.Runtime/tests/System/Reflection/NullabilityInfoContextTests.cs

```diff
@@ -0,0 +1,886 @@
+            var listOfTNotNull = typeNonNull.GetField("FieldListOfT")!;
+            listNullability = nullabilityContext.Create(listOfTNotNull);
+            Assert.Equal(NullabilityState.Nullable, listNullability.TypeArguments[0].ReadState);
+            Assert.Equal(typeof(string), listNullability.TypeArguments[0].Type);
+
+            var dictStringToTNotNull = typeNonNull.GetField("FieldDictionaryStringToT")!;
+            dictNullability = nullabilityContext.Create(dictStringToTNotNull);
+            Assert.Equal(NullabilityState.NotNull, dictNullability.TypeArguments[0].ReadState);
+            Assert.Equal(NullabilityState.Nullable, dictNullability.TypeArguments[1].ReadState);
+            Assert.Equal(typeof(string), dictNullability.TypeArguments[1].Type);
+
+            var typeOpen = typeof(GenericTest<>);
+            var listOfTOpen = typeOpen.GetField("FieldListOfT")!;
+            listNullability = nullabilityContext.Create(listOfTOpen);
+            Assert.Equal(NullabilityState.Nullable, listNullability.TypeArguments[0].ReadState);
+            // Assert.Equal(typeof(T), listNullability.TypeArguments[0].Type);
+
+            var dictStringToTOpen = typeOpen.GetField("FieldDictionaryStringToT")!;
+            dictNullability = nullabilityContext.Create(dictStringToTOpen);
+            Assert.Equal(NullabilityState.NotNull, dictNullability.TypeArguments[0].ReadState);
+            Assert.Equal(NullabilityState.Nullable, dictNullability.TypeArguments[1].ReadState);
+        }
+
+        public static IEnumerable<object[]> MethodReturnParameterTestData()
+        {
+            yield return new object[] { "MethodReturnsUnknown", NullabilityState.Unknown, NullabilityState.Unknown};
+            yield return new object[] { "MethodReturnsNullNon", NullabilityState.Nullable, NullabilityState.NotNull };
+            yield return new object[] { "MethodReturnsNullNull", NullabilityState.Nullable, NullabilityState.Nullable };
+            yield return new object[] { "MethodReturnsNonNull", NullabilityState.NotNull, NullabilityState.Nullable };
+            yield return new object[] { "MethodReturnsNonNotNull", NullabilityState.NotNull, NullabilityState.NotNull };
+            yield return new object[] { "MethodReturnsNonMaybeNull", NullabilityState.NotNull, NullabilityState.Nullable };
+            yield return new object[] { "MethodReturnsNonNon", NullabilityState.NotNull, NullabilityState.NotNull };
+        }
+
+        [Theory]
+        [MemberData(nameof(MethodReturnParameterTestData))]
+        public void MethodReturnParameterTest(string methodName, NullabilityState elementState, NullabilityState readState)
+        {
+            var method = testType.GetMethod(methodName, flags)!;
+            var nullability = nullabilityContext.Create(method.ReturnParameter);
+            Assert.Equal(readState, nullability.ReadState);
+            //Assert.Equal(readState, nullability.WriteState);
```

Comment:
Not sure if the method return parameter's read and write state should be the same, because it is evaluated as a normal parameter read and write state become different when the method has additional nullability attributes like `MaybeNull` used.

Reply:
IMO we should return the "unknown" value for cases where it can be ambiguous.

It would also be good to add tests for C# "ref returns" feature where the return value can be "writable" here; I think it should just be "unknown" for that case as well.

Reply:
> It would also be good to add tests for C# "ref returns" feature where the return value can be "writable" here; I think it should just be "unknown" for that case as well.

Wonder why you think WriteState should be "unknown" where the return value can be "writable" for ref returns, anyway i will add a test for that

Reply:
> Wonder why you think WriteState should be "unknown" where the return value can be "writable" for ref returns, anyway i will add a test for that

I'm OK with "unknown" for ref returns as long as it is consistent with other usages of `ref` in normal parameters.

Reply:
> I'm OK with "WriteState" for ref returns (and not "unknown") as long as it is consistent with other usages of ref in normal parameters.

For now, any return parameters evaluated the same as normal parameters, my question was if we should treat the return parameters differently, so sounds like we want to keep them same as normal parameters.

> IMO we should return the "unknown" value for cases where it can be ambiguous.

Another ambiguous case is a read-only properties, should we set the write state for them as `Unknown` cc @terrajobst